driver-core/src/main/com/mongodb/internal/client/model/BulkWriteBatch.java

```diff
@@ -0,0 +1,391 @@
+        this.command = command;
+        this.payload = payload;
+        this.unprocessed = unprocessed;
+    }
+
+    public void addResult(final BsonDocument result) {
+        if (writeConcern.isAcknowledged()) {
+            if (hasError(result)) {
+                MongoBulkWriteException bulkWriteException = getBulkWriteException(result);
+                bulkWriteBatchCombiner.addErrorResult(bulkWriteException, indexMap);
+            } else {
+                bulkWriteBatchCombiner.addResult(getBulkWriteResult(result), indexMap);
+            }
+        }
+    }
+
+    public BsonDocument getCommand() {
+        return command;
+    }
+
+    public SplittablePayload getPayload() {
+        return payload;
+    }
+
+    public Decoder<BsonDocument> getDecoder() {
+        return DECODER;
+    }
+
+    public BulkWriteResult getResult() {
+        return bulkWriteBatchCombiner.getResult();
+    }
+
+    public boolean hasErrors() {
+        return bulkWriteBatchCombiner.hasErrors();
+    }
+
+    public MongoBulkWriteException getError() {
+        return bulkWriteBatchCombiner.getError();
+    }
+
+    public boolean shouldProcessBatch() {
+        return !bulkWriteBatchCombiner.shouldStopSendingMoreBatches() && !payload.isEmpty();
+    }
+
+    public BulkWriteBatch getNextBatch() {
+        if (payload.hasAnotherSplit()) {
+            IndexMap nextIndexMap = IndexMap.create();
+            int newIndex = 0;
+            for (int i = payload.getPosition(); i < payload.getPayload().size(); i++) {
+                nextIndexMap = nextIndexMap.add(newIndex, indexMap.map(i));
+                newIndex++;
+            }
+
+            return new BulkWriteBatch(namespace, serverAddress, ordered, writeConcern, bypassDocumentValidation, bulkWriteBatchCombiner,
+                    nextIndexMap, batchType, command, payload.getNextSplit(), unprocessed);
+        } else {
+            return new BulkWriteBatch(namespace, serverAddress, ordered, writeConcern, bypassDocumentValidation, bulkWriteBatchCombiner,
+                    unprocessed);
+        }
+    }
+
+    private FieldNameValidator getFieldNameValidator() {
+        if (batchType == INSERT) {
+            return new CollectibleDocumentFieldNameValidator();
+        } else if (batchType == UPDATE || batchType == REPLACE) {
+            Map<String, FieldNameValidator> rootMap = new HashMap<String, FieldNameValidator>();
+            if (batchType == WriteRequest.Type.REPLACE) {
+                rootMap.put("u", new CollectibleDocumentFieldNameValidator());
+            } else {
+                rootMap.put("u", new UpdateFieldNameValidator());
+            }
+            return new MappedFieldNameValidator(new NoOpFieldNameValidator(), rootMap);
+        } else {
+            return new NoOpFieldNameValidator();
+        }
+    }
+
+    private BsonDocument createAndValidatePayloadItem(final WriteRequest writeRequest) {
+        OutputBuffer buffer = new BasicOutputBuffer();
```

Comment:
Rather than just wrapping the document in a BsonDocumentWrapper we create a RawBsonDocument so that we can validate the WriteRequest.